pulser-simulation/pulser_simulation/simresults.py

```diff
@@ -524,25 +524,32 @@ def sample_state(
         if self._meas_errors is None:
             return sampled_state
 
-        detected_sample_dict: Counter = Counter()
-        for shot, n_detects in sampled_state.items():
-            eps = self._meas_errors["epsilon"]
-            eps_p = self._meas_errors["epsilon_prime"]
-            # Shot as an array of 1s and 0s
-            shot_arr = np.array(list(shot), dtype=int)
-            # Probability of flipping each bit
-            flip_probs = np.array([eps_p if x == "1" else eps for x in shot])
-            # 1 if it flips, 0 if it stays the same
-            flips = (
-                np.random.uniform(size=(n_detects, len(flip_probs)))
-                < flip_probs
-            ).astype(int)
-            # XOR betwen the original array and the flips
-            # Gives an array of n_detects individual shots
-            new_shots = shot_arr ^ flips
-            # Count all the new_shots
-            detected_sample_dict += Counter(
-                "".join(map(str, measured)) for measured in new_shots
-            )
+        eps = self._meas_errors["epsilon"]
+        eps_p = self._meas_errors["epsilon_prime"]
+
+        if eps_p == 0.0 and eps == 0.0:
+            return sampled_state
 
-        return detected_sample_dict
+        shots = list(sampled_state.keys())
+        n_detects_list = list(sampled_state.values())
+
+        # Convert shots to a 2D array
+        shot_arr = np.array([list(shot) for shot in shots], dtype=int)
+        # Compute flip probabilities
+        flip_probs = np.where(shot_arr == 1, eps_p, eps)
+        # Repeat flip_probs based on n_detects_list
+        flip_probs_repeated = np.repeat(flip_probs, n_detects_list, axis=0)
+        # Generate random matrix of shape (sum(n_detects_list), len(shot))
+        random_matrix = np.random.uniform(
+            size=(np.sum(n_detects_list), len(shot_arr[0]))
+        )
+        # Compare random matrix with flip probabilities
+        flips = random_matrix < flip_probs_repeated
+        # Perform XOR between original array and flips
+        new_shots = shot_arr.repeat(n_detects_list, axis=0) ^ flips
+        # Count all the new_shots
+        detected_sample_dict: Counter = Counter(map(tuple, new_shots))
+
+        return Counter(
+            {"".join(map(str, k)): v for k, v in detected_sample_dict.items()}
+        )
```

Comment:
Is this preferable over simply converting directly to an iterable of strings and giving it to `Counter`?

Reply:
Something like this:
```suggestion
        return Counter(map("".join, new_shots))
```

Reply:
I did this because it is more efficient; this is with your code modified like this:

```python
return Counter(map(lambda x: "".join(map(str, x)), new_shots))
```

```
Samples_per_run: 10	Elapsed: 4.808044672012329 seconds
Samples_per_run: 100	Elapsed: 9.669185876846313 seconds
Samples_per_run: 500	Elapsed: 33.64236116409302 seconds
```

Instead of :
```
Samples_per_run: 1	Elapsed: 4.041079044342041 seconds
Samples_per_run: 10	Elapsed: 4.325520038604736 seconds
Samples_per_run: 100	Elapsed: 6.078380107879639 seconds
Samples_per_run: 500	Elapsed: 13.140138626098633 seconds
```

Reply:
> Right, you have to convert to `str`, I forgot about that. What about (this is my final suggestion, I promise!):

Bad as well; the bootleneck here is the indexing: tuple indexing is faster than str one (don't worry :D)

```
Samples_per_run: 10	Elapsed: 5.543439865112305 seconds
Samples_per_run: 100	Elapsed: 12.317859172821045 seconds
Samples_per_run: 500	Elapsed: 35.426414489746094 seconds
```

Reply:
Ok, you win 😅 
Maybe just leave a comment explaining why you did it like this because I think at first glance it looks redundant